### PR TITLE
Add experimental NetBSD host support

### PR DESCRIPTION
**Pull request by krytarowski**
This module can already boot FREEDOS on NetBSD/amd64.

### CODE REVIEW

**Comment by krytarowski** [comment]
http://netbsd.org/~kamil/screenfetch/haxm.png

**Comment by krytarowski** [comment]
It's not a final version, but base for further improvements, in particular NetBSD as guest still does not build.

**Comment by raphaelning** [comment]
Thanks! First of all, we need you to commit to the maintenance of all the NetBSD-specific code that this PR adds to the HAXM source tree. Basically, the conditions for merging #108 (see https://github.com/intel/haxm/pull/108#issuecomment-430194073) also apply to this PR.

We don't really have time to set up a NetBSD environment and test this PR. Still, if you write up some documentation (`docs/manual-netbsd.md`), other NetBSD users may be able to try this out.

Please allow me more time to complete the review. I'm not familiar with the BSD API, so if other people want to participate in the review (@tuxillo maybe?), please feel free to do so :-)

**Comment by tuxillo** [comment]
BSDs are quite different at this level and I'm not really familiar with NetBSD internals. We decided not to share code initially because of this. If there are ports to other BSDs then we might decide to create some common layer but as @krytarowski pointed out, it shouldn't be the focus at this early stage.

I'll keep working on the DragonFly BSD port, and I'll submit it when it is more or less ready with the lessons learned from this PR :-)

**Comment by krytarowski** [comment]
CC: @m00nbsd - NetBSD x86 maintainer (for review of code, if there is time and interest -- for the discussion of rationale of HAXM - we already discussed internally).

These exact internals are almost completely distinct and specific to NetBSD. We are probably not more similar to other BSDs (FreeBSD, DragonFlyBSD) than to Linux or Darwin. OpenBSD is a fork of NetBSD and uses UVM (with 2 decades of divergence), but it doesn't support loadable kernel modules anyway.

1. I'm not touching any generic code paths of other systems. All builds pass.
2. https://github.com/NetBSD/pkgsrc-wip/tree/master/qemu-haxm patch for adding Linux to qemu is pending upstream. I will follow up with NetBSD (add a 1-liner to configure).
3. OK. I've also marked myself a maintainer also in pkgsrc-wip/haxm (for NetBSD host).

This support isn't complete, but good enough for start. I'm using locally in particular this patch for fpudna:
http://netbsd.org/~kamil/patch-0074-permit-fpudna-in-kernel.txt I'm also using a little bit patched kernel in other parts -- not sure how relevant, but I will keep researching patches one after another - starting with this FPU handling. @m00nbsd rejected the fpudna kernel patch and requested to fix HAXM, I will find a way for it - very likely it will touch generic code paths.

I propose to reschedule `docs/manual-netbsd.md` until we will be able to boot into a functional shell of NetBSD on a NetBSD host. I will keep working on it and I explicitly state that this patch adds experimental support, already functional for booting FREEDOS prompt/installer.

**Comment by krytarowski** [comment]
As a reviewer maybe CC: @zoulasc

**Comment by raphaelning** [comment]
> If possible I would like to drop !CONFIG_HAX_EPT2 from haxm after this commit. It will simplify code paths and need for all OS-specific APIs implemented in HAXM.

Yes, that would be great. There's no need for new code to support the `!CONFIG_HAX_EPT2` (legacy memory virtualization engine) code path.

**Comment by krytarowski** [comment]
> We don't really have time to set up a NetBSD environment and test this PR.

I recommend to revisit this statement. I'm going to prepare a command-by-command tutorial how to get it done so you (and maybe @AlexAltea ) can give it a try.

It will involve steps not applicable for future end-users (as patching the kernel).

**Comment by krytarowski** [comment]
Tutorial: http://netbsd.org/~kamil/haxm/haxm-tutorial.txt

**Comment by AlexAltea** [comment]
@krytarowski Thank you, I'll try to replicate your results this weekend and let you know if it works for me.

**Comment by krytarowski** [comment]
@AlexAltea thanks! I'm looking forward to it.

**Comment by krytarowski** [comment]
ping?

**Comment by raphaelning** [comment]
@krytarowski Sorry, I wasn't able to make much progress last week. I'll try to complete this review today.

File: platforms/netbsd/hax_entry_vcpu.c [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    .d_mmap = nommap,
+    .d_kqfilter = nokqfilter,
+    .d_discard = nodiscard,
+    .d_flag = D_OTHER
+};
+
+/* VCPU operations */
+
+int hax_vcpu_open(dev_t self, int flag __unused, int mode __unused,
+                         struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vcpu.c]: nit:

```suggestion
                  struct lwp *l __unused)
```

File: platforms/netbsd/components.c [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+#include <sys/param.h>
+#include <sys/types.h>
+#include <sys/device.h>
+#include <sys/kmem.h>
+
+#include "../../core/include/hax_core_interface.h"
+#include "../../core/include/config.h"
+
+extern struct cfdriver hax_vm_cd;
```
Comment (raphaelning) [inline platforms/netbsd/components.c]: Just curious: in which file is this variable actually declared? All declarations I can find are modified by `extern`.

File: platforms/netbsd/hax_entry.c [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+}
+
+static int
+hax_vcpu_detach(device_t self, int flags)
+{
+    struct hax_vcpu_softc *sc;
+
+    sc = device_private(self);
+    if (sc == NULL) {
+        hax_error("device_private() for hax_vm failed\n");
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry.c]: nit:

```suggestion
        hax_error("device_private() for hax_vcpu failed\n");
```

File: platforms/netbsd/hax_entry_vcpu.c [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+    ret = hax_vcpu_core_open(cvcpu);
+    if (ret)
+        hax_error("Failed to open core vcpu\n");
+    hax_put_vcpu(cvcpu);
+    return ret;
+}
+
+int hax_vcpu_close(dev_t self, int flag __unused, int mode __unused,
+           struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vcpu.c]: nit:

```suggestion
                   struct lwp *l __unused)
```

File: platforms/netbsd/hax_entry_vcpu.c [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    /* put the one for vcpu create */
+    hax_put_vcpu(cvcpu);
+    /* put the one just held */
+    hax_put_vcpu(cvcpu);
+
+    return 0;
+}
+
+int hax_vcpu_ioctl(dev_t self, u_long cmd, void *data, int flag,
+           struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vcpu.c]: nit:

```suggestion
                   struct lwp *l __unused)
```

File: platforms/netbsd/hax_entry_vm.c [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    .d_tty = notty,
+    .d_poll = nopoll,
+    .d_mmap = nommap,
+    .d_kqfilter = nokqfilter,
+    .d_discard = nodiscard,
+    .d_flag = D_OTHER
+};
+
+int hax_vm_open(dev_t self, int flag __unused, int mode __unused,
+          struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vm.c]: nit:

```suggestion
                struct lwp *l __unused)
```

File: platforms/netbsd/hax_entry_vm.c [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        return ENODEV;
+
+    ret = hax_vm_core_open(cvm);
+    hax_put_vm(cvm);
+    hax_log_level(HAX_LOGI, "Open VM%02d\n", vm->id);
+    return ret;
+}
+
+int hax_vm_close(dev_t self __unused, int flag __unused, int mode __unused,
+           struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vm.c]: nit:

```suggestion
                 struct lwp *l __unused)
```

File: platforms/netbsd/hax_entry_vm.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    if (cvm) {
+        /* put the ref get just now */
+        hax_put_vm(cvm);
+        hax_put_vm(cvm);
+    }
+    return 0;
+}
+
+int hax_vm_ioctl(dev_t self __unused, u_long cmd, void *data, int flag,
+           struct lwp *l __unused)
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vm.c]: nit:

```suggestion
                 struct lwp *l __unused)
```

File: platforms/netbsd/hax_entry_vm.c [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    return ret;
+}
+
+int hax_vm_close(dev_t self __unused, int flag __unused, int mode __unused,
+           struct lwp *l __unused)
+{
+    struct hax_vm_softc *sc;
+    struct vm_t *cvm;
+    struct hax_vm_netbsd_t *vm;
+    int ret;
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vm.c]: Remove this unused variable

File: platforms/netbsd/hax_entry_vcpu.c [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    if (ret)
+        hax_error("Failed to open core vcpu\n");
+    hax_put_vcpu(cvcpu);
+    return ret;
+}
+
+int hax_vcpu_close(dev_t self, int flag __unused, int mode __unused,
+           struct lwp *l __unused)
+{
+    int ret;
```
Comment (raphaelning) [inline platforms/netbsd/hax_entry_vcpu.c]: Remove this unused variable

File: platforms/netbsd/hax_entry.c [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+            goto init_err8;
+        }
+        err = devsw_attach(HAX_VCPU_DEVICE_NAME, NULL, &hax_vcpu_bmajor, &hax_vcpu_cdevsw,
+                       &hax_vcpu_cmajor);
+        if (err) {
+            hax_error("Failed to register HAXM VCPU device\n");
+            goto init_err9;
+        }
+
+        for (i = 0; i < HAX_MAX_VMS; i++)
```
[the remark this answers is not in the document]
Comment (krytarowski) [reply platforms/netbsd/hax_entry.c]: NetBSD does not support dynamic devfs so I'm creating all the devices entries on module load.

We need to create the file nodes manually (mknod(8)).

I'm manipulating whether each one is available through `device_private` entry.

This looks different to others, but if this is considered a hack, what would be the expected solution?

File: platforms/netbsd/components.c (hunk 2 above)
Comment (krytarowski) [reply platforms/netbsd/components.c]: https://nxr.netbsd.org/xref/src/sys/sys/device.h#355

```
    375 #define	CFDRIVER_DECL(name, class, attrs)				\
    376 struct cfdriver __CONCAT(name,_cd) = {					\
    377 	.cd_name		= ___STRING(name),			\
    378 	.cd_class		= class,				\
    379 	.cd_attrs		= attrs,				\
    380 }
```

```
platforms/netbsd/hax_entry.c:CFDRIVER_DECL(hax_vm, DV_DULL, hax_vm_attrs);
platforms/netbsd/hax_entry.c:CFDRIVER_DECL(hax_vcpu, DV_DULL, hax_vcpu_attrs);
```

File: platforms/netbsd/hax_entry_hax.c [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        break;
+    }
+    case HAX_IOCTL_CREATE_VM: {
+        int vm_id;
+        struct vm_t *cvm;
+
+        cvm = hax_create_vm(&vm_id);
+        if (!cvm) {
+            hax_log_level(HAX_LOGE, "Failed to create the HAX VM\n");
+            ret = ENOMEM;
```
[the remark this answers is not in the document]
Comment (krytarowski) [reply platforms/netbsd/hax_entry_hax.c]: BSD uses positive numbers, I've noted that qemu expects negative ones.. so I will fix it.

File: platforms/netbsd/hax_entry.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (raphaelning) [reply platforms/netbsd/hax_entry.c]: > This looks different to others, but if this is considered a hack, what would be the expected solution?

I don't know. Does NetBSD impose a limit on the number of device nodes? Now that HAX_MAX_VMS == 64, we need to 64 x 16 nodes upfront. Is that a lot of overhead, in terms of memory and time consumption?

Anyway, I just thought adding a TODO comment would be helpful.

[the remark this answers is not in the document]
Comment (krytarowski) [reply platforms/netbsd/hax_entry.c]: 64x16 is a lot of devices - I will check whether there is a limit.

I don't have hardware to support more than the 8 cpu threads, so inflating limit doesn't win anything on my end right now, and there are more important points to put my focus (namely ability to run OSes correctly).

File: include/netbsd/hax_interface_netbsd.h [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+};
+
+struct hax_vcpu_netbsd_t;
+
+struct hax_vcpu_softc {
+    device_t sc_dev;
+    struct hax_vcpu_netbsd_t *vcpu;
+};
+
+#define HAX_MAX_VMS 8
```
[the remark this answers is not in the document]
Comment (krytarowski) [reply include/netbsd/hax_interface_netbsd.h]: Fixed

File: platforms/netbsd/hax_entry.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (krytarowski) [reply platforms/netbsd/hax_entry.c]: I've decided to stay right now with 8VMs max and expand to 64 VMs in future as a low priority feature.

**Comment by krytarowski** [comment]
@raphaelning done.

I'm trying to leave refactoring of this code for future and get this merged as is.

**Comment by krytarowski** [comment]
If the new version is acceptable, I will squash it and rebase on top of master.

File: core/include/config.h [hunk 14]
```diff
@@ -66,4 +66,12 @@ struct config_t {
 
 #define HAX_MAX_VCPUS 16
 
+#ifdef __NetBSD__
```
Comment (raphaelning) [inline core/include/config.h]: nit: `core/` code should use `HAX_PLATFORM_*` which is more readable:

```suggestion
#ifdef HAX_PLATFORM_NETBSD
```

**Comment by krytarowski** [comment]
Done!

**Comment by krytarowski** [comment]
Thanks!

I've noted that it didn't take long to see HAXM in action:
https://twitter.com/sehnsuc77565442/status/1072182648125251584

**Pull request merged**